File: 547-number-of-provinces.cpp
int findCircleNum(vector<vector<int>>& isConnected) {
    int n = isConnected.size();
    int count = 0;
    vector<bool> visited(n, false);
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            dfs(isConnected, i, visited);
            ++count;
        }
    }
    return count;
}

void dfs(vector<vector<int>>& isConnected, int i, vector<bool>& visited) {
    visited[i] = true;
    for (int j = 0; j < isConnected.size(); ++j) {
        if (isConnected[i][j] == 1 && !visited[j])
            dfs(isConnected, j, visited);
    }
}

File: python version/547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        father = {}
        self.count = n
        def find(x):
            if x not in father:
                return x
            father[x] = find(father[x])
            return father[x]
        
        def merge(x, y):
            root_x = find(x)
            root_y = find(y)
            if root_x != root_y:
                father[root_x] = root_y
                self.count -= 1
        
        for i in range(n):
            for j in range(i, n):
                if isConnected[i][j] == 1:
                    merge(i, j)
        return self.count